crates/oxc_linter/src/rules/no_extra_semi.rs

```diff
@@ -0,0 +1,72 @@
+use oxc_ast::{AstKind, Span};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+
+use crate::{autofix::Fix, context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-extra-semi): Unnecessary semicolon.")]
+#[diagnostic()]
+struct NoExtraSemiDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoExtraSemi;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Checks for usage of unnecessary extra semi-colons
+    ///
+    /// ### Why is this bad?
+    /// Extra semi-colons can cause confusion when they're typically not intentional.
+    ///
+    ///
+    /// ### Example
+    /// ```javascript
+    /// function foo() {};
+    /// ```
+    NoExtraSemi
+);
+
+impl NoExtraSemi {
+    fn is_valid_parent<'a>(&self, kind: &AstKind) -> bool {
+        match kind {
+            AstKind::ForStatement(_) => true,
+            AstKind::ForInStatement(_) => true,
+            AstKind::ForOfStatement(_) => true,
+            AstKind::WhileStatement(_) => true,
+            AstKind::DoWhileStatement(_) => true,
+            AstKind::IfStatement(_) => true,
+            AstKind::LabeledStatement(_) => true,
+            AstKind::WithStatement(_) => true,
+            _ => false,
+        }
+    }
+}
+
+impl Rule for NoExtraSemi {
```

Comment:
I implemented this rule so I can test against something that is deemed "fixable"

Reply:
Ok, but let's not commit this into main. I'll explain and we'll discuss about this later.